### PR TITLE
 stylo: Allow computing change hints during the traversal.

### PR DESCRIPTION
**Pull request by emilio**
---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

- [x] These changes do not require tests because geckolib :-(

r? @bholley 
cc @heycam

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12645)

### CODE REVIEW

File: components/style/gecko_glue.rs [hunk 1]
```diff
@@ -47,6 +46,10 @@ impl<GeckoType, ServoType> ArcHelpers<GeckoType, ServoType> {
         unsafe { transmute(owned) }
     }
 
+    pub unsafe fn borrow(borrowed: &Arc<ServoType>) -> *const &mut GeckoType {
```
Comment (bholley) [inline components/style/gecko_glue.rs]: Can we make this a closure-y API like |with| to avoid the unsafety here?

**Comment by bholley** [comment]
Generally seems reasonable, but will need some fiddling in light of https://bugzilla.mozilla.org/show_bug.cgi?id=1290335#c9

File: components/style/dom.rs [hunk 2]
```diff
@@ -46,7 +46,23 @@ impl OpaqueNode {
 }
 
 pub trait TRestyleDamage : BitOr<Output=Self> + Copy {
-    fn compute(old: Option<&Arc<ComputedValues>>, new: &ComputedValues) -> Self;
+    /// The source for our current computed values in the cascade. This is a
+    /// ComputedValues in Servo and a StyleContext in Gecko.
+    ///
+    /// This must be obtained via the node.
+    type CascadeStyleSource;
```
Comment (bholley) [inline components/style/dom.rs]: Let's call this PreExistingComputedValues?

File: components/style/dom.rs [hunk 3]
```diff
@@ -159,6 +175,13 @@ pub trait TNode : Sized + Copy + Clone {
     fn unstyle(self) {
         self.mutate_data().unwrap().style = None;
     }
+
+    /// XXX: It's a bit unfortunate we need to pass the current computed values
+    /// as an argument here, but otherwise Servo would crash due to double
+    /// borrows to return it.
+    fn style_source<'a>(&'a self,
```
Comment (bholley) [inline components/style/dom.rs]: existing_style_for_restyle_damage?

File: components/style/animation.rs [hunk 4]
```diff
@@ -507,7 +507,8 @@ where Damage: TRestyleDamage {
                                                                  frame);
             if updated_style {
                 if let Some(damage) = damage {
-                    *damage = *damage | Damage::compute(Some(style), &new_style);
+                    *damage =
+                        *damage | Damage::compute_for_layout(Some(style), &new_style);
```
Comment (bholley) [inline components/style/animation.rs]: Let's return the new style and let the caller handle it (or, if that doesn't work, let the caller provide a callback that gets passed the new style), and then remove the distinction between compute_for_layout and compute_for_cascade.

File: components/style/gecko_glue.rs [hunk 5]
```diff
@@ -47,6 +46,15 @@ impl<GeckoType, ServoType> ArcHelpers<GeckoType, ServoType> {
         unsafe { transmute(owned) }
     }
 
+    pub fn borrow<F, Output>(borrowed: &Arc<ServoType>, cb: F) -> Output
```
Comment (bholley) [inline components/style/gecko_glue.rs]: with_gecko?

File: ports/geckolib/wrapper.rs [hunk 6]
```diff
@@ -277,6 +313,21 @@ impl<'ln> TNode for GeckoNode<'ln> {
...
+        use std::mem;
+
+        if current_cv.is_none() {
+            // Don't bother in doing an ffi call to get null back.
+            return None;
+        }
+
+        unsafe {
+            let context_ptr = Gecko_GetStyleContext(self.node);
+            mem::transmute(context_ptr)
```
Comment (bholley) [inline ports/geckolib/wrapper.rs]: unsafe { context_ptr.as_ref() } ?

File: components/style/dom.rs (hunk 2 above)
Comment (bholley) [inline components/style/dom.rs]: Also, please add a comment explaining why we do this dance with the style context (i.e. the inference from the caching).

**Comment by bholley** [comment]
@bors-servo delegate+

r=me with comments.

**Comment by emilio** [comment]
@bholley: I'll need you to review the last commit and the regen.py changes in the previous one.

File: ports/geckolib/sanity_checks.rs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    }
+}
+
+#[test]
+fn assert_restyle_hints_match() {
+    use style::restyle_hints::*; // For flags
+    use gecko_bindings::structs::nsRestyleHint;
+
+    check_enum_value_non_static!(nsRestyleHint::eRestyle_Self, RESTYLE_SELF.bits());
+    // XXX This for servo actually means something like an hipotetical
```
Comment (bholley) [inline ports/geckolib/sanity_checks.rs]: "a hypothetical"

File: ports/geckolib/sanity_checks.rs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    // Restyle_AllDescendants (but without running selector matching on the
+    // element). ServoRestyleManager interprets it like that, but in practice we
+    // should align the behavior with Gecko adding a new restyle hint, maybe?
```
Comment (bholley) [inline ports/geckolib/sanity_checks.rs]: Reference the bug we have on file for doing that?

**Comment by bholley** [comment]
lgtm

**Pull request merged**